firedrake/linear_solver.py

```diff
@@ -131,21 +139,17 @@ def solve(self, x, b):
         if not isinstance(b, function.Function):
             raise TypeError("Provided RHS is a '%s', not a Function" % type(b).__name__)
 
-        if len(self._W) > 1 and self.nullspace is not None:
-            self.nullspace._apply(self._W.dof_dset.field_ises)
-        if len(self._W) > 1 and self.transpose_nullspace is not None:
-            self.transpose_nullspace._apply(self._W.dof_dset.field_ises, transpose=True)
-        if len(self._W) > 1 and self.near_nullspace is not None:
-            self.near_nullspace._apply(self._W.dof_dset.field_ises, near=True)
+        if len(self.trial_space) > 1 and self.nullspace is not None:
+            self.nullspace._apply(self.trial_space.dof_dset.field_ises)
+        if len(self.test_space) > 1 and self.transpose_nullspace is not None:
+            self.transpose_nullspace._apply(self.test_space.dof_dset.field_ises,
+                                            transpose=True)
+        if len(self.trial_space) > 1 and self.near_nullspace is not None:
+            self.near_nullspace._apply(self.trial_space.dof_dset.field_ises, near=True)
+
```

Comment:
So now that both the test and trial spaces are recorded, `solve` should do the right thing for Petrov-Galerkin discretizations with bcs, right? Would it be worthwhile to add a small test solving a simple PDE using a PG-method with bcs?

Reply:
No, at least not for general PG I think.  Think about what that means: the test and trial spaces are different.

The current interface only allows providing the bc for the row (trial) space.  For PG, you need to identify which nodes are killed in the test space given boundary conditions on the trial space.  But this is not a trivial identification (for straight Galerkin, it's easy, because the nodes are the same).

Now, I admit that I have not thought hard about what the maths (and therefore the code) looks like for this case.  

c.f. also the discussion in #197.

Reply:
I thought that with PG one always had a structural identity between the test and trial space, with only the basis function values being different between test and trial.